### ISSUE TITLE
Running brew update results in process hang / freeze

### ISSUE THREAD

**Issue by chanonroy** [Homebrew/brew#14187]
### `brew doctor` output

```shell
Your system is ready to brew.
```

### Verification

- [X] My "`brew doctor` output" above says `Your system is ready to brew.` and am still able to reproduce my issue.
- [X] I ran `brew update` twice and am still able to reproduce my issue.
- [X] This issue's title and/or description do not reference a single formula e.g. `brew install wget`. If they do, open an issue at https://github.com/Homebrew/homebrew-core/issues/new/choose instead.

### `brew config` output

```shell
HOMEBREW_VERSION: 3.6.12-22-gd98de74
ORIGIN: https://github.com/Homebrew/brew
HEAD: d98de7438a9d6efb8049d0f1d59109b98f7a1d7b
Last commit: 29 minutes ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 3b2af9221dda2be65b3319f2e066875de4c7b734
Core tap last commit: 2 weeks ago
Core tap branch: master
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 8
Homebrew Ruby: 2.6.10 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: octa-core 64-bit arm_firestorm_icestorm
Clang: 14.0.0 build 1400
Git: 2.38.1 => /opt/homebrew/bin/git
Curl: 7.84.0 => /usr/bin/curl
macOS: 13.0.1-arm64
CLT: 14.1.0.0.1.1666437224
Xcode: 14.1
Rosetta 2: false
```

### What were you trying to do (and why)?

I was trying to update brew with `brew update`

### What happened (include all command output)?

While trying to run `brew update`, the system hanged.

I then ran `brew update -d` to run in debug mode and noticed that it hung on the following logs:

```
rm -f /opt/homebrew/.git/TMP_FETCH_FAILURES
```

### What did you expect to happen?

I expected brew to update

### Step-by-step reproduction instructions (by running `brew` commands)

```shell
> brew update
> brew update -d
```

**Comment by trusche** [comment]
Same issue. adding my config here in case it helps:

```
HOMEBREW_VERSION: 3.6.3
ORIGIN: https://github.com/Homebrew/brew
HEAD: d5b6b068c15a8df60f0f5e34a4ca80dcf82c48e3
Last commit: 9 weeks ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 018dea9ea239093f4b613c9b7aaf5339a9224109
Core tap last commit: 9 weeks ago
Core tap branch: master
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CASK_OPTS: []
HOMEBREW_EDITOR: vim
HOMEBREW_MAKE_JOBS: 6
Homebrew Ruby: 2.6.8 => /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.8_1/bin/ruby
CPU: hexa-core 64-bit kabylake
Clang: 14.0.0 build 1400
Git: 2.37.3 => /usr/local/bin/git
Curl: 7.79.1 => /usr/bin/curl
macOS: 12.6-x86_64
CLT: 14.1.0.0.1.1666437224
Xcode: 14.1
```

**Comment by trusche** [comment]
Running `brew update-reset` before `brew update` fixed this for me.

**Comment by chanonroy** [comment]
> Running brew update-reset before brew update fixed this for me.
This also fixed it for me 🎉

**Comment by brandonryan** [comment]
Same issue here, but `brew update-reset` is hanging on homebrew-core for me too...

**Comment by MikeMcQuaid** [comment]
When you say "hanging"/"freezing" here: how long are you waiting? Is your internet connection (and specifically: connection to GitHub) otherwise stable/fast?

**Comment by chanonroy** [comment]
> When you say "hanging"/"freezing" here: how long are you waiting? 

After waiting for at least 5 minutes, I've always killed the process, so I haven't seen how far the hanging/freezing lasts.

> Is your internet connection (and specifically: connection to GitHub) otherwise stable/fast?

Yes. GitHub connection is fine and internet speeds at 100mb/s.

> Running `brew update-reset` before `brew update` fixed this for me.
As mentioned above, this was what resolved it for me. It resolved quickly as well.

**Comment by Bo98** [comment]
We've seen some performance regressions recently with fetching and pushing to homebrew-core, which I assume is the same issue described here.

It's not consistent - it sometimes takes the usual time but sometimes can hang for a long time.

It sounds like something on GitHub's side in any case. I'm not sure if it's some partially rolled out change or what.

**Comment by andygeorge** [comment]
Running into this issue as well on an initial fresh install, with no indication of how to fix.

**Comment by MikeMcQuaid** [comment]
> _Running_ into this issue as well on an initial fresh install, with no indication of how to fix.

There's nothing we can really do to fix. We're not GitHub. We don't control your connection to GitHub.

**Comment by crazy4groovy** [comment]
I'm also getting an error on fresh install:

```
stevenolsen@MacBook-Air github % /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"

==> Checking for `sudo` access (which may request your password)...
Password:
==> This script will install:
/opt/homebrew/bin/brew
/opt/homebrew/share/doc/homebrew
/opt/homebrew/share/man/man1/brew.1
/opt/homebrew/share/zsh/site-functions/_brew
/opt/homebrew/etc/bash_completion.d/brew
/opt/homebrew

Press RETURN/ENTER to continue or any other key to abort:
==> /usr/bin/sudo /usr/sbin/chown -R stevenolsen:admin /opt/homebrew
==> Downloading and installing Homebrew...
HEAD is now at aea140c04 Merge pull request #14191 from Homebrew/sponsors-maintainers-man-completions
error: Not a valid ref: refs/remotes/origin/master
fatal: ambiguous argument 'refs/remotes/origin/master': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'
error: RPC failed; curl 56 LibreSSL SSL_read: error:02FFF036:system library:func(4095):Connection reset by peer, errno 54
error: 5149 bytes of body are still expected
fetch-pack: unexpected disconnect while reading sideband packet
fatal: early EOF
fatal: fetch-pack: invalid index-pack output
Error: Fetching /opt/homebrew/Library/Taps/homebrew/homebrew-core failed!
fatal: 'origin/master' is not a commit and a branch 'master' cannot be created from it
fatal: invalid upstream 'origin/master'
Failed during: /opt/homebrew/bin/brew update --force --quiet
```

Brand new mac book air 2020 M1, trying to set up.

**Comment by bitnovus** [comment]
Running `git status` in `/usr/local/Homebrew` previously showed `On branch stable`. It appears there is no longer a `stable` branch on this repo.

After running `brew update-reset`, as suggested by @trusche, `git status` showed `On branch master` and `brew update` started working again for me.

**Comment by Bo98** [comment]
For issues with _initial_ installs, refer to https://github.com/orgs/Homebrew/discussions/666. Initial install does take several minutes and we're aware some networking setups (ISPs, proxies, etc.) have issues with that (e.g. Virgin Media).

This issue is mostly about `brew update` being suddenly being slow on _existing_ install, which is an issue I've seen less today so it might be fixed.

**Comment by DaveKT** [comment]
Getting the same issue. Running brew update just hangs (5 mins at least). Ran Brew update-reset also had a hang when Fetching /opt/homebrew/Library/Taps/homebrew/homebrew-core... but it did (after several mins resolve and proceed normally. Tested network connections between my local computer and host and am able to connect @ 1.1Gbps with no packet loss. Could still be on GitHub side but just thought I'd share my experience. Snippet of the update-reset below with the slow step bolded.

<pre>
DTassara@Gauss ~ % brew update-reset
==> Fetching /opt/homebrew...
remote: Enumerating objects: 104, done.
remote: Counting objects: 100% (76/76), done.
remote: Total 104 (delta 76), reused 76 (delta 76), pack-reused 28
Receiving objects: 100% (104/104), 65.19 KiB | 10.86 MiB/s, done.
Resolving deltas: 100% (82/82), completed with 56 local objects.
From https://github.com/Homebrew/brew
 * [new branch]          dependabot/bundler/Library/Homebrew/rubocop-1.38.0 -> origin/dependabot/bundler/Library/Homebrew/rubocop-1.38.0
 * [new branch]          dependabot/bundler/Library/Homebrew/rubocop-ast-1.24.0 -> origin/dependabot/bundler/Library/Homebrew/rubocop-ast-1.24.0
 * [new branch]          less-agressive-core-tap -> origin/less-agressive-core-tap

==> Resetting /opt/homebrew...
branch 'master' set up to track 'origin/master'.
Switched to and reset branch 'master'
Your branch is up to date with 'origin/master'.

==> Fetching /opt/homebrew/Library/Taps/homebrew/homebrew-cask...

==> Resetting /opt/homebrew/Library/Taps/homebrew/homebrew-cask...
branch 'master' set up to track 'origin/master'.
Reset branch 'master'
Your branch is up to date with 'origin/master'.

<b>==> Fetching /opt/homebrew/Library/Taps/homebrew/homebrew-core...</b> <i>Very slow at this step >5 mins</i>
remote: Enumerating objects: 25621, done.
remote: Counting objects: 100% (11790/11790), done.
remote: Compressing objects: 100% (229/229), done.
remote: Total 25621 (delta 11572), reused 11773 (delta 11561), pack-reused 13831
Receiving objects: 100% (25621/25621), 10.92 MiB | 28.75 MiB/s, done.
Resolving deltas: 100% (19071/19071), completed with 4173 local objects.
From https://github.com/Homebrew/homebrew-core
   82d49a08630..f0312b42414  master       -> origin/master
 * [new branch]              ubuntu-18.04 -> origin/ubuntu-18.04

==> Resetting /opt/homebrew/Library/Taps/homebrew/homebrew-core...
branch 'master' set up to track 'origin/master'.
Reset branch 'master'

</pre>

**Comment by 2called-chaos** [comment]
Not 100% sure if this is related but I had very slow installs (setup a new mac so installed a bunch). Eventually I figured out that this is somehow apparently related to ipv6. I copied a curl command from debug output and ran it manually with `-v` and noticed that sometimes it resolved to an v4 and sometimes to an v6 address.

When using v6 it's very slow for some reason. Sometimes it hangs just a few seconds, sometimes 5 minutes until it times out: https://gist.github.com/2called-chaos/93ff56a9162e90becad31b36273c4b41#file-gistfile1-txt-L9

Since I couldn't find a way to make brew append `-4` to curl commands I disabled v6 in network settings to install my stuff without getting mad.
